### DISCUSSION TITLE
Merge class and class:list

### DISCUSSION THREAD

**Discussion by Calvin-LL** [withastro/roadmap#629] [Proposal]
### Body

# Summary

Add `class:list` features to `class` and deprecate `class:list`.

See https://github.com/withastro/astro/pull/7858 for the implementation PR.

# Background & Motivation

I wanted a custom component to accept both `class` and `class:list` and pass down the parent's style scope, but I ended up having to use a lot of boilerplate.

See [this CodeSandbox](https://codesandbox.io/p/sandbox/silly-curie-4fds5h?file=%2Fsrc%2Fpages%2Findex.astro%3A1%2C1) for a simple example.

Initially I was also unsure as to whether `class:list` gets normalized into a string and passed down as `class`. [This maintainer seems to expect it to be the case,](https://github.com/withastro/astro/issues/7747#issuecomment-1645676781) but after digging into the code, I'm not convinced if that was ever the case.

`class:list` is old, it was first proposed in [RFC#782](https://github.com/withastro/astro/issues/782) (see the proposal for more discussions). It stands out among the other astro directives as it is the only one that modifies an element's attribute. (besides `define:vars` which only applies to `<style>` elements)

See https://github.com/withastro/astro/issues/7747 for more discussions around `class:list`.

# Example

`CompB.astro`

```astro
---
const { "class": className } = Astro.props;
---
<div class={["c2", className]} ></div>

<style>
  .c2 {
    color: red;
  }
</style>
```

`CompA.astro`

```astro
---
import CompB from "./CompB.astro";

const { small } = Astro.props;
---
<CompB class={["c1", { small }]} />

<style>
  .c1 {
    background-color: blue;
  }

  .small {
    font-size: 0.5rem;
  }
</style>
```

`<CompA />` is rendered as:

```html
<div class="c1 c2 astro-33M6635F astro-7N4XSKOT"></div>
```

# Response to Potential Arguments Against

- **Why not just normalize `class:list` into `class` before passing down to the child?**
  It could break existing component like this:

  `Button.astro`

  ```astro
  ---
  const { "class:list": classList } = Astro.props;
  ---
  <button class:list={[classList]}>
    <slot />
  </button>

  ```

  `Comp.astro`

  ```astro
  ---
  import Button from "./Button.astro";
  ---
  <Button class:list={["c1"]} />
  ```

- **A `class` prop of a component doesn't necessarily mean CSS class**
  This assumption is already broken by style scoping. When a `class` props is passed down to a component, a scoping class is automatically added. See [this page in the docs](https://docs.astro.build/en/guides/styling/#passing-a-class-to-a-child-component) and [this CodeSandbox](https://codesandbox.io/p/sandbox/gracious-flower-4hms3h?file=%2Fsrc%2Fpages%2Findex.astro%3A8%2C1) for an example.

- **This breaks the direct relationship between `class` in an element with the HTML output**
  That relationship is already broken by `style`.

**Comment by prescience-data** [comment]
I like this implementation, as it should already be fairly anti-pattern to use such a critical javascript keyword `class` as a prop name, and most IDEs will give you hell if you try to destructure this in any case:

![image](https://github.com/withastro/roadmap/assets/65471523/1142be27-5527-439a-b438-c18b2c806916)

If implemented, I would also suggest adding usage of "class" as a prop name as a hard error in https://github.com/ota-meshi/eslint-plugin-astro

**Reply by Calvin-LL** [reply]
Hmm I'm not suggesting any changes to the `class` prop name. Are you proposing that Astro should move to something other than `class` for CSS classes? Not sure I understand what you want to change it to.

I have been doing [this from the docs](https://docs.astro.build/en/guides/styling/#passing-a-class-to-a-child-component):

```astro
---
const { class: className } = Astro.props;
---
<div class={className}>
  <slot/>
</div>
```

**Reply by prescience-data** [reply]
Ah sorry, I might have (overly optimistically) read into this that "class" would be automatically applied to the custom component rather than always needing to do this name reassignment and manually application.

### Example
```astro
<!-- MyComponent.astro -->
<!-- Remove requirement to always add boilerplate "class" handling code and automatically apply + merge "class" -->
<div class="pt-1">
  <slot />
</div>
```

```astro
---
import MyComponent from "./MyComponent.astro"
---
<div>
	<!-- <div class="pt-1">foo</div>	-->
	<!-- No additional classes applied-->
	<MyComponent>
		foo
	</MyComponent>

	<!-- <div class="pt-1 mt-1">foo</div>	-->
	<!-- Add `mt-1` to existing component classes -->
	<MyComponent class="mt-1">
		foo
	</MyComponent>
	
	<!-- <div class="pt-1 mt-1">foo</div> -->
	<!-- Merge `class:list` behaviour, and deduplicate existing "pt-1" class -->
	<MyComponent class={["mt-1", "pt-1"]}>
		foo
	</MyComponent>
</div>
```

Edit: As I typed this out I can see cases where you might want to modify or filter the values in the incoming `class` prop, rather than be forced always accept it as is, so perhaps the behaviour might be to auto-merge _unless specifically hijacked in `Astro.props`_, in which case the developer is responsible for their own merge / apply strategy?

**Reply by Calvin-LL** [reply]
As much as I like the _auto-merge unless specifically hijacked_ idea I think that might be a bit too magical for Astro. Also that would break stuff so maybe better left as a suggestion for v3.

I have edited the example to make the proposal more clear. Sorry for the confusion.

**Comment by mayank99** [comment]
Thank you for this! I've always found it difficult to work with `class:list` when composing components, and I've seen the behavior randomly change across minor versions of Astro multiple times. Intentionally normalizing this pattern would be quite nice.

I also appreciate the thought given to this part in particular:
> **This breaks the direct relationship between class in an element with the HTML output**
> That relationship is already broken by style.

---

Couple questions:
1. Would it make sense to deprecate `class:list` in v2 and remove it in v3? Or would that be too disruptive?
2. Should classes always be automatically merged? Or always require explicit merging? Below, the first example is already correctly handled, but the second example requires some boilerplate to correctly handle.
   ```astro
   <p {...Astro.props}>
   <style>
     p { ... }
   </style>
   ```
   ```astro
   <p class='foo' {...Astro.props}>
   <style>
     .foo { ... }
   </style>
   ```

**Reply by Calvin-LL** [reply]
> Would it make sense to deprecate class:list in v2 and remove it in v3? Or would that be too disruptive?

I think it's worth removing in v3. It wouldn't be that much work for users to do a replace all "class:list" to "class" when migrating. Plus then it'll be one fewer thing to maintain.

> Should classes always be automatically merged? Or always require explicit merging? Below, the first example is already correctly handled, but the second example requires some boilerplate to correctly handle.

I think the second case should not be automatically merged, that would be too magical for Astro and adds a special case for props merging.

**Comment by natemoo-re** [comment]
I just left [a comment on the original issue](https://github.com/withastro/astro/issues/7747#issuecomment-1684407885) that spawned this proposal, which I highly recommend everyone read.

`class:list` was never intended to be exposed as a prop to components, but this behavior was introduced during a refactor and our test suite didn't catch it because the eventual output was still correct. We're switching back to the original behavior (`class:list` should always be merged into `class`) in 3.0.

Definitely happy to discuss if `class:list` is even necessary for 4.0, but we didn't want to rush it in under the wire. Generally, we've made a conscious effort to avoid "magic" props that don't have a namespace, hence why we introduced this as `class:list` rather than overloading the plain `class` attribute.
